usr.sbin/portsnap/portsnap/portsnap.8
.\"-
.\" Copyright 2004-2005 Colin Percival
.\" Copyright 2018 Kris Moore
.\" All rights reserved
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted providing that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
.\" DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
.\" STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
.\" IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.\" $FreeBSD$
.\"
.Dd July 1, 2018
.Dt PORTSNAP 8
.Os FreeBSD
.Sh NAME
.Nm portsnap
.Nd Extract and update snapshots of the ports tree from git
.Sh SYNOPSIS
.Nm
.Op Fl I
.Op Fl b Ar gitbranch
.Op Fl f Ar conffile
.Op Fl p Ar portsdir
.Op Fl r Ar gitremote
.Cm command ...
.Op Ar path
.Sh DESCRIPTION
The
.Nm
tool is used to extract and update
the
.Fx
ports tree.
.Pp
In a normal update operation,
.Nm
will routinely restore modified files to their unmodified state and
delete unrecognized local files.
.Sh OPTIONS
The following options are supported:
.Bl -tag -width "-f conffile"
.It Fl f Ar conffile
Read the configuration from
.Ar conffile .
(default:
.Pa /etc/portsnap.conf )
.It Fl p Ar portsdir
When extracting or updating an uncompressed snapshot,
operate on the directory
.Ar portsdir .
(default:
.Pa /usr/ports/ ,
or as given in the configuration file.)
.It Fl r Ar gitremote
Pull ports tree from specified git remote.
(default: https://github.com/trueos/trueos-ports, or as given in the
configuration file.)
.Sh COMMANDS
The
.Cm command
can be any one of the following:
.Bl -tag -width "-f conffile"
.Cm cron
command.
.It cron
Sleep a random amount of time between 1 and 3600 seconds,
then operate as if the
.Cm update
command was specified.
As the name suggests, this command is designed for running
from
.Xr cron 8 ;
the random delay serves to minimize the probability that
a large number of machines will simultaneously attempt to
fetch updates.
.It extract
Extract a ports tree, replacing existing files and directories.
NOTE: This will remove anything occupying the location where
files or directories are being extracted; in particular, any
changes made locally to the ports tree (for example, adding new
patches) will be silently obliterated.
.Pp
Only run this command to initialize your portsnap-maintained
ports tree for the first time, if you wish to start over with
a clean, completely unmodified tree, or if you wish to extract
a specific part of the tree (using the
.Ar path
option).
.It update
Update a ports tree extracted using the
.Cm extract
command.
You must run this command to apply changes to your ports tree
after downloading updates via the
.Cm cron
commands.
Again, note that in the parts of the ports tree which are being
updated, any local changes or additions will be removed.
.It auto
Run
.Cm cron
depending on whether stdin is a terminal; then run
.Cm update
or
.Cm extract
depending on whether
.Ar portsdir
exists.
.El
.Sh TIPS
.Bl -bullet
.It
If your clock is set to local time, adding the line
.Pp
.Dl 0 3 * * * root /usr/sbin/portsnap cron
.Pp
to
.Pa /etc/crontab
is a good way to make sure you always have
an up-to-date snapshot of the ports tree available which
can quickly be extracted into
.Pa /usr/ports .
If your clock is set to UTC, please pick a random time other
than 3AM, to avoid overly imposing an uneven load on the
server(s) hosting the snapshots.
.Sh FILES
.Bl -tag -width "/etc/portsnap.conf"
.It Pa /etc/portsnap.conf
Default location of the portsnap configuration file.
.It Pa /var/db/portsnap
Default location where compressed snapshots are stored.
.It Pa /usr/ports
Default location where the ports tree is extracted.
.El
.Sh EXAMPLES
Fetch the repo and create the
.Xr ports 7
tree under
.Pa /usr/ports :
.Dl Nm Ar extract
.Pp
Update the ports tree:
.Dl Nm Ar update
.Sh SEE ALSO
.Xr fetch 1 ,
.Xr sha256 1 ,
.Xr fetch 3 ,
.Xr portsnap.conf 5 ,
.Xr pkg 7 ,
.Xr pkg-version 8
.Sh AUTHORS
.An Colin Percival Aq Mt cperciva@FreeBSD.org
.An Kris Moore Aq Mt kmoore@FreeBSD.org


usr.sbin/portsnap/portsnap/portsnap.sh
#!/bin/sh

#-
# SPDX-License-Identifier: BSD-2-Clause-FreeBSD
#
# Copyright 2004-2005 Colin Percival
# Copyright 2018 Kris Moore
# All rights reserved
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted providing that the following conditions
# are met:
# 1. Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
# 2. Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in the
#    documentation and/or other materials provided with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
# IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
# DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
# OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
# STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
# IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.

# $FreeBSD$

usage() {
	cat <<EOF
usage: `basename $0` [options] command ... [path]

Options:
  -b branch    -- Git branch to use from remote
                  (default: trueos-master)
  -f conffile  -- Read configuration options from conffile
                  (default: /etc/portsnap.conf)
  -p portsdir  -- Location of uncompressed ports tree
                  (default: /usr/ports/)
  -r remote    -- Git Server from which to fetch updates.
                  (default: https://github.com/trueos/trueos-ports)
Commands:
  cron         -- Sleep rand(3600) seconds, and then extract updates.
  extract      -- Extract snapshot of ports tree, replacing existing
                  files and directories.
  update       -- Update ports tree to match current snapshot, replacing
                  files and directories which have changed.
  auto         -- Fetch updates, and either extract a new ports tree or
                  update an existing tree.
EOF
	exit 0
}

#### Parameter handling functions.

# Initialize parameters to null, just in case they're
# set in the environment.
init_params() {
	PORTSDIR=""
	CONFFILE=""
	COMMAND=""
	COMMANDS=""
	GITREMOTE="https://github.com/trueos/trueos-ports"
	GITBRANCH="trueos-master"
}

# Parse the command line
parse_cmdline() {
	while [ $# -gt 0 ]; do
		case "$1" in
		-b)
			if [ $# -eq 1 ]; then usage; fi
			if [ ! -z "${GITBRANCH}" ]; then usage; fi
			shift; GITBRANCH="$1"
			;;
		-f)
			if [ $# -eq 1 ]; then usage; fi
			if [ ! -z "${CONFFILE}" ]; then usage; fi
			shift; CONFFILE="$1"
			;;
		-h | --help | help)
			usage
			;;
		-p)
			if [ $# -eq 1 ]; then usage; fi
			if [ ! -z "${PORTSDIR}" ]; then usage; fi
			shift; PORTSDIR="$1"
			;;
		-r)
			if [ $# -eq 1 ]; then usage; fi
			if [ ! -z "${GITREMOTE}" ]; then usage; fi
			shift; GITREMOTE="$1"
			;;
		cron | extract | fetch | update | auto)
			COMMANDS="${COMMANDS} $1"
			;;
		up)
			COMMANDS="${COMMANDS} update"
			;;
		alfred)
			COMMANDS="${COMMANDS} auto"
			;;
		*)
			if [ $# -gt 1 ]; then usage; fi
			if echo ${COMMANDS} | grep -vq extract; then
				usage
			fi
			EXTRACTPATH="$1"
			;;
		esac
		shift
	done

	if [ -z "${COMMANDS}" ]; then
		usage
	fi
}

# If CONFFILE was specified at the command-line, make
# sure that it exists and is readable.
sanity_conffile() {
	if [ ! -z "${CONFFILE}" ] && [ ! -r "${CONFFILE}" ]; then
		echo -n "File does not exist "
		echo -n "or is not readable: "
		echo ${CONFFILE}
		exit 1
	fi
}

# If a configuration file hasn't been specified, use
# the default value (/etc/portsnap.conf)
default_conffile() {
	if [ -z "${CONFFILE}" ]; then
		CONFFILE="/etc/portsnap.conf"
	fi
}

# Read {GITREMOTE, PORTSDIR, GITBRANCH} from the configuration
# file if they haven't already been set.  If the configuration
# file doesn't exist, do nothing.
parse_conffile() {
	if [ -r "${CONFFILE}" ]; then
		for X in PORTSDIR GITREMOTE GITBRANCH; do
			eval _=\$${X}
			if [ -z "${_}" ]; then
				eval ${X}=`grep "^${X}=" "${CONFFILE}" |
				    cut -f 2- -d '=' | tail -1`
			fi
		done
	fi
}

# If parameters have not been set, use default values
default_params() {
	_PORTSDIR="/usr/ports"
	_GITREMOTE="https://github.com/trueos/trueos-ports"
	_GITBRANCH="trueos-master"
	for X in PORTSDIR GITREMOTE GITBRANCH; do
		eval _=\$${X}
		eval __=\$_${X}
		if [ -z "${_}" ]; then
			eval ${X}=${__}
		fi
	done
}

# Perform sanity checks and set some final parameters
# in preparation for extracting or updating ${PORTSDIR}
# Complain if ${PORTSDIR} exists but is not writable,
# but don't complain if ${PORTSDIR} doesn't exist.
extract_check_params() {
	_PORTSDIR_bad="Directory is not writable: "

	if [ -d "${PORTSDIR}" ] && ! [ -w "${PORTSDIR}" ]; then
		echo -n "`basename $0`: "
		echo -n "${_PORTSDIR_bad}"
		echo ${PORTSDIR}
		exit 1
	fi
}

# Perform sanity checks and set some final parameters
# in preparation for updating ${PORTSDIR}
update_check_params() {
	extract_check_params

	if ! [ -r ${PORTSDIR}/.git ]; then
		echo "${PORTSDIR} was not created by git."
		echo -n "You must run '`basename $0` extract' before "
		echo "running '`basename $0` update'."
		exit 1
	fi

}

#### Core functionality -- the actual work gets done here

# Do the actual work involved in "extract"
extract_run() {
	mkdir -p ${PORTSDIR} || return 1

	if [ -d ${PORTSDIR}/.git ] ; then
		rm -rf ${PORTSDIR}/.git
	fi

	cd ${PORTSDIR} || return 1
	git init || return 1
	git remote add origin $GITREMOTE || return 1
	git fetch --depth=1 || return 1
	git reset --hard origin/$GITBRANCH || return 1
	git checkout $GITBRANCH || return 1
	git clean -f -d || return 1
}

update_run_extract() {
	cd ${PORTSDIR} || return 1
	git fetch || return 1
	git reset --hard origin/$GITBRANCH || return 1
	git checkout $GITBRANCH || return 1
}

# Do the actual work involved in "update"
update_run() {
	update_run_extract || return 1
}

#### Main functions -- call parameter-handling and core functions

# Using the command line, configuration file, and defaults,
# set all the parameters which are needed later.
get_params() {
	init_params
	parse_cmdline $@
	sanity_conffile
	default_conffile
	parse_conffile
	default_params
}

# Cron command.  Make sure the parameters are sensible; wait
# rand(3600) seconds; then fetch updates.  While fetching updates,
# send output to a temporary file; only print that file if the
# fetching failed.
cmd_cron() {
	sleep `jot -r 1 0 3600`
}

# Extract command.  Make sure the parameters are sensible,
# then extract the ports tree (or part thereof).
cmd_extract() {
	extract_check_params
	extract_run || exit 1
}

# Update command.  Make sure the parameters are sensible,
# then update the ports tree.
cmd_update() {
	update_check_params
	update_run || exit 1
}

# Auto command. Run 'update' or
# 'extract' depending on whether ${PORTSDIR} exists.
cmd_auto() {
	if [ -r ${PORTSDIR}/.git ]; then
		cmd_update
	else
		cmd_extract
	fi
}

#### Entry point

# Make sure we find utilities from the base system
export PATH=/sbin:/bin:/usr/sbin:/usr/bin:${PATH}

# Set LC_ALL in order to avoid problems with character ranges like [A-Z].
export LC_ALL=C

which -s git
if [ $? -ne 0 ] ; then
	echo "The 'git' command was not found."
	echo "Please install with 'pkg install git' and try again"
fi

get_params $@
for COMMAND in ${COMMANDS}; do
	cmd_${COMMAND}
done
